cartridges/int_algolia/cartridge/scripts/algolia/steps/sendDeltaExportProducts.js

```diff
@@ -0,0 +1,288 @@
+
+            var catalogFile = new File(l5_catalogDir, 'catalog.xml');
+
+            // adding productsIDs from the XML to the list of changed productIDs
+            // success contains the number of successfully read records or false if an error occurred
+            success = jobHelper.updateChangedProductsObjectFromXML(catalogFile, changedProducts);
+
+            if (success) {
+                // success contains the number of successfully read records
+                productLogData.processedRecords += success;
+
+                // deleting successfully processed files and their parent folders
+                success = catalogFile.remove() && l5_catalogDir.remove();
+            } else {
+                // abort if error reading from any of the delta export zips
+                let errorMessage = 'Error reading from file: ' + filename;
+                jobHelper.logError(errorMessage);
+                productLogData.processedErrorMessage = errorMessage;
+                algoliaData.setLogData(updateLogType, productLogData);
+                return new Status(Status.ERROR);
+            }
+        });
+
+        // cleanup: removing unzipped files that are already processed, along with their parent dirs
+        // this removes `l4_catalogsDir`, `version.txt` from `l3_uuidDir`, `l3_uuidDir` and `l2_tempZipDir` itself
+        jobHelper.removeFolderRecursively(l2_tempZipDir);
+    });
+
+    // writing number of records read from the SFCC delta zips
+    jobHelper.logInfo(productLogData.processedRecords + ' records read from SFCC delta zips');
+
+
+    // cleanup - removing "_processing" dir
+    jobHelper.removeFolderRecursively(l1_processingDir);
+
+
+    // ----------------------------- PART 2: Retrieving and enriching the products, writing them to XML -----------------------------
+
+
+    // open Delta XML file to write
+    var updateFile, updateFileWriter, updateXmlWriter;
+    try {
+        // <siteID>_product_update.xml
+        updateFile = new File(l0_deltaExportDir, algoliaConstants.ALGOLIA_DELTA_EXPORT_UPDATE_FILE_NAME);
+
+        // if there's already an update XML from a previous unsuccessful attempt, remove it
+        if (updateFile.exists()) {
+            updateFile.remove();
+        }
+
+        // creating file, writing start elements
+        updateFileWriter = new FileWriter(updateFile, 'UTF-8');
+        updateXmlWriter = new XMLStreamWriter(updateFileWriter);
+        updateXmlWriter.writeStartDocument();
+        updateXmlWriter.writeStartElement('products');
+
+    } catch (error) {
+        let errorMessage = 'Error opening delta XML for writing';
+        jobHelper.logFileError(updateFile.fullPath, errorMessage, error);
+        productLogData.processedErrorMessage = errorMessage;
+        algoliaData.setLogData(updateLogType, productLogData);
+        return new Status(Status.ERROR);
+    }
+
+    // retrieving products from database and enriching them
+    for (var currentObject of changedProducts) {
```

Comment:
for of loops not supported in SFCC compatibility version 18.10. Job should run successfully on 18.10 by changing this to a standard for loop.

Reply:
@jginnardhibbett, I can see that for..of loops were added in compatibility mode 21.2, but I can't find anything about `forEach`.
Can you let me know if `forEach` is supported in compatibility mode 18.10? That one should be even faster than the other two.